patroni/postgresql/sync.py

```diff
@@ -273,71 +295,118 @@ def _process_replica_readiness(self, cluster: Cluster, replica_list: _ReplicaLis
 
-        - `maximum_lag_on_syncnode`: would help swapping unhealthy sync replica in case if it stops
-          responding (or hung). Please set the value high enough so it won't unncessarily swap sync
-          standbys during high loads. Any value less or equal of 0 keeps the behavior backward compatible.
-          Please note that it will not also swap sync standbys in case where all replicas are hung.
-        - `synchronous_node_count`: controlls how many nodes should be set as synchronous.
+            - ``maximum_lag_on_syncnode``: would help swapping unhealthy sync replica in case it stops
+              responding (or hung). Please set the value high enough, so it won't unnecessarily swap sync
+              standbys during high loads. Any value less or equal to ``0`` keeps the behavior backwards compatible.
+              Please note that it will also not swap sync standbys when all replicas are hung.
+
+            - ``synchronous_node_count``: controls how many nodes should be set as synchronous.
+
+        :param cluster: current cluster topology from DCS
 
-        :returns: tuple of candidates :class:`CaseInsensitiveSet` and synchronous standbys :class:`CaseInsensitiveSet`.
+        :returns: current synchronous replication state as a :class:`_SyncState` object
         """
         self._handle_synchronous_standby_names_change()
 
         replica_list = _ReplicaList(self._postgresql, cluster)
         self._process_replica_readiness(cluster, replica_list)
 
+        active = CaseInsensitiveSet()
+        sync_nodes = CaseInsensitiveSet()
+        numsync_confirmed = 0
+
         if TYPE_CHECKING:  # pragma: no cover
             assert self._postgresql.global_config is not None
         sync_node_count = self._postgresql.global_config.synchronous_node_count\
             if self._postgresql.supports_multiple_sync else 1
         sync_node_maxlag = self._postgresql.global_config.maximum_lag_on_syncnode
 
-        candidates = CaseInsensitiveSet()
-        sync_nodes = CaseInsensitiveSet()
         # Prefer members without nofailover tag. We are relying on the fact that sorts are guaranteed to be stable.
         for replica in sorted(replica_list, key=lambda x: x.nofailover):
             if sync_node_maxlag <= 0 or replica_list.max_lsn - replica.lsn <= sync_node_maxlag:
-                candidates.add(replica.application_name)
-                if replica.sync_state == 'sync' and replica.application_name in self._ready_replicas:
-                    sync_nodes.add(replica.application_name)
-            if len(candidates) >= sync_node_count:
-                break
-
-        return candidates, sync_nodes
-
-    def set_synchronous_standby_names(self, sync: Collection[str]) -> None:
-        """Constructs and sets "synchronous_standby_names" GUC value.
+                if self._postgresql.global_config.is_quorum_commit_mode:
+                    # add nodes with nofailover tag only to get enough "active" nodes
```

Comment:
```suggestion
                    # We do not add nodes with `nofailover` enabled because that reduces availability.
                    # We need to check LSN quorum only among nodes that are promotable because
                    # there is a chance that a non-promotable node is ahead of a promotable one.
```

Would something like this make more sense based on the thread thoughts?